File: Problems & Solutions/M/Missing number in matrix/25 December 2022/Java/Missing number in matrix.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            int[][] matrix = new int[n][n];
            for(int i = 0; i < n; i++){
                String[] S = br.readLine().trim().split(" ");
                for(int j = 0; j < n; j++)
                    matrix[i][j] = Integer.parseInt(S[j]);
            }
            Solution ob = new Solution();
            long ans = ob.MissingNo(matrix);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends

class Solution
{
    public long  MissingNo(int[][] matrix)
    {
        TreeSet<Long>nonzero=new TreeSet<>(), zero=new TreeSet<>();
    	
    	for(int i = 0; i < matrix.length; i++){
    		boolean f = true;
    		long sum = 0;
    		for(int j = 0; j < matrix[0].length; j++){
    			sum += matrix[i][j];
    			if(matrix[i][j]==0)
    				f = false;
    		}
    		if(f)nonzero.add(sum);
    		if(!f)zero.add(sum);
    	}
    	for(int  j = 0; j < matrix[0].length; j++){
    		boolean f = true;
    		long sum = 0;
    		for(int i = 0; i < matrix.length; i++){
    			sum += matrix[i][j];
    			if(matrix[i][j]==0)
    				f = false;
    		}
    		if(f)nonzero.add(sum);
    		if(!f)zero.add(sum);
    	}
    	boolean f = true;
    	long sum = 0;
    	for(int i = 0; i < matrix.length; i++){
    		sum += matrix[i][i];
    		if(matrix[i][i]==0)
    			f = false;
    	}
    	if(f)nonzero.add(sum);
		if(!f)zero.add(sum);
    	f = true;
    	sum = 0;
    	for(int i = 0, j = matrix.length - 1; j >= 0; i++,j--){
    		sum += matrix[i][j];
    		if(matrix[i][j]==0)
    			f = false;
    	}
		if(f)nonzero.add(sum);
		if(!f)zero.add(sum);
    	if(nonzero.size() == 1 && zero.size() == 1){
    		long x = nonzero.first();
    		long y = zero.first();
    		if(x <= y)
    			return -1;
    		return x - y;
    	}
    	return -1;
    }
}

File: Problems & Solutions/M/Missing number in matrix/25 December 2022/C++/Missing number in matrix.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
public:
    long long int MissingNo(vector<vector<int> >& matrix) {
    	set<long long int>nonzero, zero;
    	
    	for(int i = 0; i < matrix.size(); i++){
    		bool f = true;
    		long long int sum = 0;
    		for(int j = 0; j < matrix[0].size(); j++){
    			sum += matrix[i][j];
    			if(!matrix[i][j])
    				f = false;
    		}
    		if(f)nonzero.insert(sum);
    		if(!f)zero.insert(sum);
    	}
    	for(int  j = 0; j < matrix[0].size(); j++){
    		bool f = true;
    		long long int sum = 0;
    		for(int i = 0; i < matrix.size(); i++){
    			sum += matrix[i][j];
    			if(!matrix[i][j])
    				f = false;
    		}
    		if(f)nonzero.insert(sum);
    		if(!f)zero.insert(sum);
    	}
    	bool f = true;
    	long long int sum = 0;
    	for(int i = 0; i < matrix.size(); i++){
    		sum += matrix[i][i];
    		if(!matrix[i][i])
    			f = false;
    	}
    	if(f)nonzero.insert(sum);
		if(!f)zero.insert(sum);
    	f = true;
    	sum = 0;
    	for(int i = 0, j = matrix.size() - 1; j >= 0; i++,j--){
    		sum += matrix[i][j];
    		if(!matrix[i][j])
    			f = false;
    	}
		if(f)nonzero.insert(sum);
		if(!f)zero.insert(sum);
    	if(nonzero.size() == 1 and zero.size() == 1){
    		long long int x = *nonzero.begin();
    		long long int y = *zero.begin();
    		if(x <= y)
    			return -1;
    		return x - y;
    	}
    	return -1;
    }
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		vector<vector<int>> matrix(n, vector<int>(n,0));
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n; j++){
				cin >> matrix[i][j];
			}
		}
		Solution ob;
		long long int ans = ob.MissingNo(matrix);
		cout << ans <<"\n";
	}
	return 0;
}
// } Driver Code Ends
